File: cosmos-master/code/dynamic_programming/largest_sum_contiguous_subarray/maxSubArraySum.py
# Function to find the maximum contiguous subarray using Kadane's algorithm
def maxSubArraySum(a,size):

    max_so_far = -9223372036854775806 #negative sys.maxint 
    max_ending_here = 0

    for i in range(0, size):
        max_ending_here = max_ending_here + a[i]
        if (max_so_far < max_ending_here):
            max_so_far = max_ending_here

        if max_ending_here < 0:
            max_ending_here = 0
    return max_so_far

# Enter the number seperated by spaces
arr = [int(x) for x in input().split()]

print("Maximum contiguous sum is", maxSubArraySum(arr,len(arr)))

File: code/mathematical-algorithms/max_subarray_sum/max_subarray_sum.cpp
#include <iostream>

using namespace std;

inline int mx(int i, int j)
{
        return (i > j)? i : j;
}

int max_subarray_sum(int arr[], int length)
{
        int best = 0, sum = 0;

        for(int i = 0; i < length; ++i) {
                sum = mx(arr[i], sum + arr[i]);
                best = mx(best, sum);
        }

        return best;
}

int main()
{
        int array[] = {1, 0, -1, 3, -12, 4, 0, 1, -3};
        int n = 9;
        int best = max_subarray_sum(array, n);
        cout << best << endl;
        return 0;
}